lms/djangoapps/courseware/date_summary.py

```diff
@@ -0,0 +1,248 @@
+            return _('To earn a certificate, you must complete all requirements before this date.')
+        return _('This course is archived, which means you can review course content but it is no longer active.')
+
+    @property
+    def date(self):
+        return self.course.end
+
+
+class VerifiedUpgradeDeadlineDate(DateSummary):
+    """
+    Displays the date before which learners must upgrade to the
+    Verified track.
+    """
+    css_class = 'verified-upgrade-deadline'
+    title = _('Verification Upgrade Deadline')
+    description = _('You are still eligible to upgrade to a Verified Certificate!')
+    link_text = _('Upgrade to Verified Certificate')
+
+    @property
+    def link(self):
+        return reverse('verify_student_upgrade_and_verify', args=(self.course.id,))
+
+    @lazy
+    def date(self):
+        try:
+            verified_mode = CourseMode.objects.get(
+                course_id=self.course.id, mode_slug=CourseMode.VERIFIED
+            )
+            return verified_mode.expiration_datetime
+        except CourseMode.DoesNotExist:
+            return None
+
+
+class VerificationDeadlineDate(DateSummary):
+    """
+    Displays the date by which the user must complete the verification
+    process.
+    """
+
+    @property
+    def css_class(self):
+        base_state = 'verification-deadline'
+        if self.deadline_has_passed():
+            return base_state + '-passed'
+        elif self.must_retry():
+            return base_state + '-retry'
+        else:
+            return base_state + '-upcoming'
+
+    @property
+    def link_text(self):
+        return self.link_table[self.css_class][0]
+
+    @property
+    def link(self):
+        return self.link_table[self.css_class][1]
+
+    @property
+    def link_table(self):
+        """Maps verification state to a tuple of link text and location."""
+        return {
+            'verification-deadline-passed': (_('Learn More'), ''),
+            'verification-deadline-retry': (_('Retry Verification'), reverse('verify_student_reverify')),
+            'verification-deadline-upcoming': (
+                _('Verify My Identity'),
+                reverse('verify_student_verify_now', args=(self.course.id,))
+            )
+        }
+
+    @property
+    def title(self):
+        if self.deadline_has_passed():
+            return _('Missed Verification Deadline')
+        return _('Verification Deadline')
+
+    @property
+    def description(self):
+        if self.deadline_has_passed():
+            return _(
+                "Unfortunately you missed this course's deadline for"
+                " a successful verification."
+            )
+        return _(
+            "You must successfully complete verification before"
+            " this date to qualify for a Verified Certificate."
+        )
+
+    @lazy
+    def date(self):
+        return VerificationDeadline.deadline_for_course(self.course.id)
+
+    @property
+    def is_enabled(self):
```

Comment:
Should this be `@lazy` as well since `CourseEnrollment.enrollment_mode_for_user` makes a call to the db?